contracts/Donations.sol

```diff
@@ -0,0 +1,193 @@
+    }
+
+    /**
+     * Transfers the donated funds in the currency @param _token to the charity with the id @param _destinationId.
+     *
+     * @param _destinationId ID of the charity.
+     * @param _token Currency to transfer the funds from.
+     * @param _to Address of the charity.
+     */
+    function donate(
+        uint256 _destinationId,
+        IERC20 _token,
+        address _to
+    ) external onlyRole(WORKER_ROLE) {
+        require(
+            transferableAmounts[_token][_destinationId] != 0,
+            "Donations: nothing to donate"
+        );
+
+        uint256 amount = transferableAmounts[_token][_destinationId];
+        transferableAmounts[_token][_destinationId] = 0;
+
+        _token.safeTransfer(_to, amount);
+
+        emit DonationsSent(_destinationId, _token, _to, amount);
+    }
+
+    /**
+     * Mints an NFT for every donation in @param _params.
+     * The @param _groupId is used to uniquely identify this group of donations.
+     * Ideally, @param _groupId is the hash of the transaction where the yield for the donations was claimed to by the treasury.
+     *
+     * @param _groupId Unique identifier for the group of donations in @param _params.
+     * @param _params Donation params.
+     */
+    function mint(bytes32 _groupId, DonationParams[] calldata _params)
+        external
+        onlyRole(WORKER_ROLE)
+    {
+        require(
+            !processedDonationsGroups[_groupId],
+            "Donations: already processed"
+        );
+
+        uint64 expiry = _getBlockTimestamp() + ttl;
+        uint256 length = _params.length;
+
+        for (uint256 i = 0; i < length; i++) {
+            uint256 _metadataId = metadataId.current();
+            metadataId.increment();
+
+            metadata[_metadataId] = Metadata({
+                destinationId: _params[i].destinationId,
+                token: _params[i].token,
+                expiry: expiry,
+                amount: _params[i].amount
+            });
+
+            _mint(_params[i].owner, _metadataId);
+
+            emit DonationMinted(
+                _metadataId,
+                _params[i].destinationId,
+                _groupId,
+                _params[i].token,
+                expiry,
+                _params[i].amount,
+                _params[i].owner
+            );
+        }
+
+        processedDonationsGroups[_groupId] = true;
+    }
+
+    /**
+     * Burns the NFT and sets the amount donated to be transferred to the charity.
+     *
+     * @param _id ID of the NFT.
+     */
+    function burn(uint256 _id) external {
+        require(
+            ownerOf(_id) == _msgSender() || hasRole(WORKER_ROLE, _msgSender()),
+            "Donations: not allowed"
+        );
+
+        uint256 destinationId = metadata[_id].destinationId;
+        IERC20 token = metadata[_id].token;
+        uint256 amount = metadata[_id].amount;
+
+        transferableAmounts[token][destinationId] += amount;
+
+        _burn(_id);
```

Comment:
shouldn't `metadata[_id].expiry` come into play here somehow?

Reply:
oh, yes! I forgot! The worker can only burn after the expiry!